Short link info and generic goldenlayout

partouf opened the pull request:
For the mobile app I'm making, I wondered if I could get it to open CE shortlinks, but shortlink is stored in its original goldenlayout format, which has way way way more info than I would like to receive.

So the first thing this PR adds is:
* /api/shortlinkinfo/:id - translates goldenlayout to basic client state and returns in JSON format

And for fun, because Matt was talking about being able to store state without having to depend on GL, I made:
* /resetlayout/:id - rebuilds a new goldenlayout from the basic client state that was derived from the original goldenlayout, so same code and compilers, different clean layout (so this is able to fix broken /z/ links)

Offers a workaround for https://github.com/mattgodbolt/compiler-explorer/issues/1077 and other goldenlayout issues.

**Todo now**
- [x] API for Apps to load clientstate from links
- [x] Better name for /y/ -> now is /resetlayout/
- [x] Option to make shortlinks from clientstate

**Todo later**
- Add optional layout to clientstate
- Able to interpret goldenlayout to clientstate layout
- Able to regenerate goldenlayout from clientstate layout

mattgodbolt commented:
Thanks @partouf ! Will take a look at this tomorrow!

RabsRincon commented on lib/clientstate-normalizer.js:
```diff
@@ -0,0 +1,345 @@
+                    },
+                    isClosable: true,
+                    reorderEnabled: true,
+                    title: compiler.id + " Opt Viewer (Editor #" + session.id + ", Compiler #" + compilerIndex
```
The title is regenerated clientside using the ids on its state, so you're right, there should be no need to duplicate it here

partouf replied:
yeah the title is regenerated, it was really only for debugging, ill remove it

partouf commented on app.js:
```diff
@@ -483,6 +571,8 @@ aws.initConfig(awsProps)
                     .use('/api', apiHandler.handle)
                     .use('/g', oldGoogleUrlHandler)
                     .get('/z/:id', storedStateHandler)
+                    .get('/y/:id', storedStateHandlerRebuilt)
```
agree with shortlinkinfo, should be in api, was planning on doing that but lazy/forgot

partouf commented on lib/clientstate-normalizer.js:
```diff
@@ -0,0 +1,345 @@
+
+    fromClientState(state) {
+        this.golden = {
+            settings: {
```
I just copy/pasted a freshly reset state, and then it worked, so I have no idea what it all means

partouf commented on lib/clientstate.js:
```diff
@@ -0,0 +1,110 @@
+// POSSIBILITY OF SUCH DAMAGE.
+"use strict";
+
+class ClientStateCompilerOptions {
```
It's true that there's no layout information. I had some ideas, but not entirely sure yet what the best way is to approach that.

partouf commented:
Finally figured out what to do with the /y/:id link. I renamed it to /resetlayout/:id and added an option in the dropdown menu for it when you're at a /z/ link page.

![Reset link](https://i.imgur.com/5ko8DIe.png)
